### PR TITLE
Reimplement MPCD solvent forces

### LINKED ISSUE #663: MPCD and GPU polymorphism
## Description

The MPCD forces are currently implemented with runtime polymorphism. This works well on the CPU but is complicated on the GPU because it requires a special wrapper to construct/destroy the force object and also needs separable device compilation. @jglaser has pointed that this polymorphism does not work on AMD GPUs. Another difficulty is how plugins can be built against hoomd.mpcd with either new streaming geometries or new forces, since they no longer have access to the unlinked device objects. We should try to figure out a better solution to this problem.

## Proposed solution

Two possibilities:

1. Supply the MPCD CUDA objects unlinked as a static library (in addition to the MPCD shared library), and plugins can then complete the device link themselves. I tried unsuccessfully to get this to work in older CMake, but it may be easier with modern CMake. This doesn't fix the AMD issue, though.

2. Scrap the runtime polymorphism and template the streaming methods on the force in addition to the geometry. This will increase compile time and forces developers to template all geometries when they add a new force, but will certainly work.

There may also be changes required for how integrators, streaming methods, and forces work for the v3 API, so I'm not sure if one of these might 

## Developer

Yes, I can look into this when I refactor for v3.

### PR DESCRIPTION
**Pull request by mphoward**
## Description

This PR adds back the ability to apply a body force on the solvent. The force is a class that is attached to a streaming method.

This PR also removes the complicated machinery that was in place to do GPU polymorphism, which barely worked with CUDA and did not work with ROCm. Instead, classes are now templated.

This PR should be merged after #1691.

## Motivation and context

Solvent forces are an important feature for people using MPCD. This also fixes a long-time issue with polymorphism that prevented hoomd.mpcd from compiling for AMD GPUs.

Resolves #663

## How has this been tested?

The force API and streaming method are tested at the Python level. The forces themselves are tested by a reimplemented C++ test.

## Change log

N/A, but it could be noted that `GPUPolymorph` was removed. As far as I know, it has only been used in one spot in the MPCD code, and it has some pretty severe limitations with compilation.

## Checklist:

- [X] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [X] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [X] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by mphoward** [comment]
Not stale

File: hoomd/mpcd/ConstantForce.cc
Comment (tommy-waltmann) [inline hoomd/mpcd/ConstantForce.cc:34]: I would prefer to see these defined in the `ConstantForce` class than in the export function if possible.

File: hoomd/mpcd/pytest/test_stream.py
Comment (tommy-waltmann) [inline hoomd/mpcd/pytest/test_stream.py:75]: Should a case where no force is added to the streaming method be included for completeness?

**Review by tommy-waltmann — COMMENTED** [review]
Overall, looks to be in good shape. I just have two small things I want to mention.

File: hoomd/mpcd/CMakeLists.txt
Comment (joaander) [inline hoomd/mpcd/CMakeLists.txt:120]: Thanks for following this pattern! It really helps with compile times.

**Review by joaander — APPROVED** [review]
Thanks! @tommy-waltmann has a few suggestions for improvement if you would like to implement before merging.

Have you tested mpcd on an AMD GPU?

**Review by rayasare — APPROVED** [review]

File: hoomd/mpcd/CMakeLists.txt
Comment (mphoward) [reply hoomd/mpcd/CMakeLists.txt:120]: Yes definitely, and it is also way more convenient for generating the boilerplate code for the different templates!

File: hoomd/mpcd/ConstantForce.cc
Comment (mphoward) [reply hoomd/mpcd/ConstantForce.cc:34]: Since these are short functions that are only used for binding and would need to be conditionally compiled in the class, I prefer to keep them here.

File: hoomd/mpcd/pytest/test_stream.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_stream.py:75]: I think this case is covered by `test_pickling` because `solvent_force=None` is the default, but it doesn't hurt to explicitly test with `None` here as well. Added!

**Comment by mphoward** [comment]
I applied Tommy's suggestion on the unit tests and replied on the other one, so this should be ready to merge!
 
> Have you tested mpcd on an AMD GPU?

I unfortunately don't have access to an AMD GPU. I removed Jens' comment about MPCD not working with AMD GPUs because I remember that issue was related to device polymorphism, which is removed in this PR. (Jens had added CMake code to stop compilation of MPCD for AMD with HIP.)

**Review by tommy-waltmann — APPROVED** [review]

**Comment by mphoward** [comment]
@joaander unless you want to hold off due to the comment about AMD above, could you merge this when you get a chance? Thanks!

**Comment by joaander** [comment]
I will build the mpcd-v4 branch and run tests on Frontier when I get a chance. I would like to do that before merging mpcd-v4 into a trunk branch.

**Comment by mphoward** [comment]
That would be great, thanks! If possible, let's do that as the final step before merging as there may be one or two more C++ refactorings if I have time.

I'll add this to the checklist in #775 .

**Pull request merged**
